src/alignment_algorithms.jl

```diff
@@ -0,0 +1,148 @@
+# flat doubly stochastic matrix, starting point
+"""
+    _flat_doubly_stochastic(n)
+
+Given the dimension ``n``, return the flat doubly stochastic matrix ``J=\\frac{1}{n}\\mathbb{1}\\mathbb{1}^T``.
+The flat doubly stochastic matrix is the barycenter of doubly stochastic matrices.
+"""
+_flat_doubly_stochastic(n::Int64) = ones(n) * ones(n)' / n
+
+# gradient of the objective function
+"""
+    _gradient(A,B,P_i)
+
+Given the adjacency matrices ``A`` and ``B``, return the gradient evaluated in ``P_i`` of the function ``f(P)=-\\operatorname{trace}(APB^TP^T)``.
+"""
+function _gradient(
+    A::AbstractMatrix{U}, B::AbstractMatrix{V}, P_i::AbstractMatrix{T}
+) where {U<:Real,V<:Real,T<:Real}
+    return -A * P_i * B' - A' * P_i * B
+end
+
+# norm
+"""
+    _distance(A,B,P)
+
+Given the adjacency matrices ``A`` and ``B`` and the permutation matrix ``P``, compute the distance between the permuted graphs.
+"""
+function _distance(
+    A::AbstractMatrix{U}, B::AbstractMatrix{V}, P::AbstractMatrix{T}
+) where {U<:Real,V<:Real,T<:Real}
+    return norm(A * P - P * B)
+end
+
+# solve assignment problem
+"""
+    _solve_assignment_problem(data; optimizer)
+
+Given a data matrix and a JuMP optimizer, return the integer solution of the assignment problem.
+"""
+function _solve_assignment_problem(data::AbstractMatrix{U}; optimizer) where {U<:Real}
+    M, N = size(data)
+    model = Model(optimizer)
+    set_silent(model)
+    @variable(model, x[m in 1:M, n in 1:N] >= zero(U), integer = true)
+    @objective(model, Max, sum(data[m, n] * x[m, n] for n in 1:N, m in 1:M),)
+    @constraint(model, [m in 1:M], sum(x[m, :]) == one(U))
+    @constraint(model, [n in 1:N], sum(x[:, n]) == one(U))
+    optimize!(model)
+    P = zeros(M, N)
+    for n in 1:N
+        for m in 1:M
+            P[m, n] = value(x[m, n])
+        end
+    end
+    return P
+end
+
+"""
+    _step_size(A,B,P_i,Q_i)
+
+Given ``A,B,P_i`` and the direction matrix ``Q_i``, return the step size of the gradient descent method.
+"""
+function _step_size(
+    A::AbstractMatrix{U},
+    B::AbstractMatrix{V},
+    P_i::AbstractMatrix{T},
+    Q_i::AbstractMatrix{W},
+) where {U<:Real,V<:Real,T<:Real,W<:Real}
+    R = P_i - Q_i
+    a = -tr(A * R * B' * R')
+    b = -tr(A * Q_i * B' * R' + A * R * B' * Q_i')
+    if a > 0.0
+        α = min(max(-0.5 * b / a, 0.0), 1.0)
+    else
+        if a + b > 0.0
+            α = 0.0
```

Comment:
Does `α` have to be a `Float64`?

Reply:
Yes, even if we start the algorithm with `α` integer, it will become a float with the iterations.